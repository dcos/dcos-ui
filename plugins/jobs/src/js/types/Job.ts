import {
  JobDetailResponse as MetronomeJobDetailResponse,
  GenericJobResponse as MetronmeGenericJobResponse,
  JobResponse as MetronomeJobResponse,
  JobStatus as MetronomeJobStatus,
  isDetailResponse as isMetronomeJobDetailResponse
} from "#SRC/js/events/MetronomeClient";
import {
  JobRunConnection,
  JobRunConnectionTypeResolver,
  JobRunConnectionSchema
} from "#PLUGINS/jobs/src/js/types/JobRunConnection";
import {
  JobHistorySummary,
  JobHistorySummaryTypeResolver,
  JobHistorySummarySchema
} from "#PLUGINS/jobs/src/js/types/JobHistorySummary";
import {
  JobRunStatusSummary,
  JobRunStatusSummaryTypeResolver,
  JobRunStatusSummarySchema
} from "#PLUGINS/jobs/src/js/types/JobRunStatusSummary";
import {
  JobScheduleConnection,
  JobScheduleConnectionTypeResolver,
  JobScheduleConnectionSchema
} from "#PLUGINS/jobs/src/js/types/JobScheduleConnection";
import {
  JobStatus,
  JobStatusSchema
} from "#PLUGINS/jobs/src/js/types/JobStatus";
import { JobHistoryRun } from "#PLUGINS/jobs/src/js/types/JobHistoryRun";
import {
  JobDockerTypeResolver,
  JobDocker,
  JobDockerSchema
} from "#PLUGINS/jobs/src/js/types/JobDocker";
import { JobLabel, LabelSchema } from "#PLUGINS/jobs/src/js/types/JobLabel";
import { cleanJobJSON } from "#SRC/js/utils/CleanJSONUtil";

export interface Job {
  activeRuns: JobRunConnection | null;
  command: string;
  cpus: number;
  description: string | null;
  disk: number;
  docker: JobDocker | null;
  id: string;
  jobRuns: JobRunConnection | null;
  json: string;
  labels: JobLabel[];
  lastRunsSummary: JobHistorySummary;
  lastRunStatus: JobRunStatusSummary;
  mem: number;
  name: string;
  path: string[];
  schedules: JobScheduleConnection;
  scheduleStatus: JobStatus;
}

export const JobSchema = `
${JobRunConnectionSchema}
${LabelSchema}
${JobHistorySummarySchema}
${JobRunStatusSummarySchema}
${JobStatusSchema}
${JobDockerSchema}
${JobScheduleConnectionSchema}

type Job {
  activeRuns: JobRunConnection
  command: String!
  cpus: Float!
  description: String
  disk: Float!
  docker: Docker
  id: ID!
  jobRuns: JobRunConnection
  json: String!
  labels: [Label]!
  lastRunsSummary: JobHistorySummary
  lastRunStatus: JobRunStatusSummary
  mem: Int!
  name: String!
  path: [String]!
  schedules: ScheduleConnection!
  scheduleStatus: JobStatus!
}
`;

export function JobTypeResolver(response: MetronmeGenericJobResponse): Job {
  return {
    id: JobFieldResolvers.id(response),
    description: JobFieldResolvers.description(response),
    command: JobFieldResolvers.command(response),
    disk: JobFieldResolvers.disk(response),
    mem: JobFieldResolvers.mem(response),
    cpus: JobFieldResolvers.cpus(response),
    docker: JobFieldResolvers.docker(response),
    json: JobFieldResolvers.json(response),
    labels: JobFieldResolvers.labels(response),
    lastRunStatus: JobFieldResolvers.lastRunStatus(response),
    name: JobFieldResolvers.name(response),
    path: JobFieldResolvers.path(response),
    jobRuns: JobFieldResolvers.jobRuns(response),
    lastRunsSummary: JobFieldResolvers.lastRunsSummary(response),
    scheduleStatus: JobFieldResolvers.scheduleStatus(response),
    activeRuns: JobFieldResolvers.activeRuns(response),
    schedules: JobFieldResolvers.schedules(response)
  };
}

export const JobFieldResolvers = {
  id(job: MetronmeGenericJobResponse): string {
    return job.id;
  },
  description(job: MetronmeGenericJobResponse): string | null {
    return isMetronomeJobDetailResponse(job) ? job.description : null;
  },
  command(job: MetronmeGenericJobResponse): string {
    return job.run.cmd;
  },
  disk(job: MetronmeGenericJobResponse): number {
    return job.run.disk;
  },
  mem(job: MetronmeGenericJobResponse): number {
    return job.run.mem;
  },
  cpus(job: MetronmeGenericJobResponse): number {
    return job.run.cpus;
  },
  activeRuns(job: MetronmeGenericJobResponse): JobRunConnection | null {
    return isMetronomeJobDetailResponse(job)
      ? JobRunConnectionTypeResolver(job.activeRuns)
      : null;
  },
  docker(job: MetronmeGenericJobResponse): JobDocker | null {
    if (isMetronomeJobDetailResponse(job) && job.run.docker) {
      return JobDockerTypeResolver(job.run.docker);
    }
    return null;
  },
  jobRuns(job: MetronmeGenericJobResponse): JobRunConnection | null {
    return isMetronomeJobDetailResponse(job)
      ? AddStatusToHistoryJobRuns(job)
      : null;
  },
  json(job: MetronmeGenericJobResponse): string {
    return JSON.stringify(cleanJobJSON(job));
  },
  labels(job: MetronmeGenericJobResponse): JobLabel[] {
    return Object.entries(job.labels).map(([key, value]) => ({ key, value }));
  },
  lastRunsSummary(job: MetronmeGenericJobResponse): JobHistorySummary {
    return isMetronomeJobDetailResponse(job)
      ? JobHistorySummaryTypeResolver(job.history)
      : JobHistorySummaryTypeResolver(
          (job as MetronomeJobResponse).historySummary
        );
  },
  lastRunStatus(job: MetronmeGenericJobResponse): JobRunStatusSummary {
    return isMetronomeJobDetailResponse(job)
      ? JobRunStatusSummaryTypeResolver(job.history)
      : JobRunStatusSummaryTypeResolver(
          (job as MetronomeJobResponse).historySummary
        );
  },
  name(job: MetronmeGenericJobResponse): string {
    return job.id.split(".").pop() || "";
  },
  path(job: MetronmeGenericJobResponse): string[] {
    return job.id.split(".").slice(0, -1);
  },
  schedules(job: MetronmeGenericJobResponse): JobScheduleConnection {
    return JobScheduleConnectionTypeResolver(job.schedules);
  },
  scheduleStatus(job: MetronmeGenericJobResponse): JobStatus {
    const scheduleConnection = JobScheduleConnectionTypeResolver(job.schedules);

    if (isMetronomeJobDetailResponse(job)) {
      const jobRunConnection = AddStatusToHistoryJobRuns(job);

      if (jobRunConnection.longestRunningActiveRun !== null) {
        return jobRunConnection.longestRunningActiveRun.status;
      }
      if (jobRunConnection.nodes.length > 0) {
        const schedule = scheduleConnection.nodes[0];

        if (schedule != null && schedule.enabled) {
          return "SCHEDULED";
        }
      }
    }

    if (scheduleConnection.nodes.length === 0) {
      return "UNSCHEDULED";
    }

    return "COMPLETED";
  }
};

function AddStatusToHistoryJobRuns(
  job: MetronomeJobDetailResponse
): JobRunConnection {
  const { successfulFinishedRuns, failedFinishedRuns } = job.history;

  const successfulFinishedRunsWithStatus: JobHistoryRun[] = successfulFinishedRuns.map(
    run => ({ ...run, status: "COMPLETED" as MetronomeJobStatus }) // TODO: investiagte why we need to cast this
  );

  const failedFinishedRunsWithStatus: JobHistoryRun[] = failedFinishedRuns.map(
    run => ({
      ...run,
      status: "FAILED" as MetronomeJobStatus
    })
  );

  return JobRunConnectionTypeResolver([
    ...job.activeRuns,
    ...successfulFinishedRunsWithStatus,
    ...failedFinishedRunsWithStatus
  ]);
}
