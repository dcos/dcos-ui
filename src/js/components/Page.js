import classNames from "classnames/dedupe";
import PropTypes from "prop-types";
import React from "react";
import createReactClass from "create-react-class";
import { StoreMixin } from "mesosphere-shared-reactjs";

import { MountService } from "foundation-ui";
import BasePageHeader from "../components/PageHeader";
import FluidGeminiScrollbar from "./FluidGeminiScrollbar";
import ScrollbarUtil from "../utils/ScrollbarUtil";
import TemplateUtil from "../utils/TemplateUtil";

const PageHeader = ({
  actions,
  addButton,
  breadcrumbs,
  supplementalContent,
  tabs,
  disabledActions
}) => {
  return (
    <BasePageHeader
      actions={actions}
      addButton={addButton}
      breadcrumbs={breadcrumbs}
      supplementalContent={supplementalContent}
      tabs={tabs}
      disabledActions={disabledActions}
    />
  );
};

TemplateUtil.defineChildren(PageHeader, {
  Breadcrumbs: BasePageHeader.Breadcrumbs,
  Actions: BasePageHeader.Actions,
  Tabs: BasePageHeader.Actions
});

PageHeader.defaultProps = {
  actions: [],
  tabs: [],
  disabledActions: false
};

PageHeader.propTypes = {
  actions: PropTypes.array,
  addButton: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.object),
    PropTypes.object
  ]),
  breadcrumbs: PropTypes.node,
  supplementalContent: PropTypes.node,
  tabs: PropTypes.array,
  disabledActions: PropTypes.bool
};

const Page = createReactClass({
  displayName: "Page",

  mixins: [StoreMixin],

  getInitialState() {
    return {
      rendered: false
    };
  },

  propTypes: {
    className: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.object,
      PropTypes.string
    ]),
    dontScroll: PropTypes.bool,
    flushBottom: PropTypes.bool,
    navigation: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    title: PropTypes.oneOfType([PropTypes.object, PropTypes.string])
  },

  componentWillMount() {
    this.store_listeners = [
      {
        name: "sidebar",
        events: ["widthChange"]
      }
    ];
  },

  componentDidMount() {
    // eslint-disable-next-line
    this.setState({
      rendered: true
    });
  },

  onSidebarStoreWidthChange() {
    ScrollbarUtil.updateWithRef(this.geminiRef);
  },

  getChildren() {
    if (this.state.rendered === true) {
      // Avoid rendering template children twice
      return TemplateUtil.filterTemplateChildren(
        this.constructor,
        this.props.children
      );
    }

    return null;
  },

  getNavigation(navigation) {
    if (!navigation) {
      return null;
    }

    return <div className="page-header-navigation">{navigation}</div>;
  },

  getPageHeader() {
    return TemplateUtil.getChildOfType(
      this.props.children,
      this.constructor.Header
    );
  },

  getContent() {
    const { dontScroll, flushBottom } = this.props;
    const contentClassSet = classNames(
      "page-body-content pod pod-tall flex",
      "flex-direction-top-to-bottom flex-item-grow-1",
      {
        "flex-item-shrink-1": dontScroll,
        "flush-bottom": flushBottom
      }
    );

    const content = <div className={contentClassSet}>{this.getChildren()}</div>;

    if (dontScroll) {
      return content;
    }

    return (
      <FluidGeminiScrollbar
        autoshow={true}
        className="page-body flex flex-direction-top-to-bottom
          flex-direction-left-to-right-screen-large flex-item-grow-1
          flex-item-shrink-1 gm-scrollbar-container-flex"
        ref={ref => (this.geminiRef = ref)}
      >
        {content}
      </FluidGeminiScrollbar>
    );
  },

  render() {
    const { className, navigation, dontScroll, title } = this.props;

    const classSet = classNames(
      "page flex flex-direction-top-to-bottom flex-item-grow-1",
      {
        "flex-item-shrink-1": dontScroll
      },
      className
    );

    return (
      <div className={classSet}>
        <MountService.Mount type="Page:TopBanner" />
        {this.getPageHeader(title, navigation)}
        {this.getContent()}
      </div>
    );
  }
});

TemplateUtil.defineChildren(Page, { Header: PageHeader });

export default Page;
export const Header = PageHeader;
