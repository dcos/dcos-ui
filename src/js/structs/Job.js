import {cleanJobJSON} from '../utils/CleanJSONUtil';
import DateUtil from '../utils/DateUtil';
import Item from './Item';
import JobRunList from './JobRunList';
import MesosConstants from '../../../plugins/services/src/js/constants/MesosConstants';

module.exports = class Job extends Item {
  getActiveRuns() {
    return new JobRunList({items: this.get('activeRuns')});
  }

  getCommand() {
    const {cmd} = this.get('run') || {};

    return cmd;
  }

  getCpus() {
    const {cpus = MesosConstants.MIN_CPUS} = this.get('run') || {};

    return cpus;
  }

  getDescription() {
    return this.get('description');
  }

  getDocker() {
    const {docker = {}} = this.get('run') || {};

    return docker;
  }

  getDisk() {
    const {disk = 0} = this.get('run') || {};

    return disk;
  }

  getId() {
    return this.get('id');
  }

  getJobRuns() {
    const activeRuns = this.get('activeRuns') || [];
    const {
      failedFinishedRuns = [],
      successfulFinishedRuns = []
    } = this.get('history') || {};

    return new JobRunList({
      items: [].concat(activeRuns, failedFinishedRuns, successfulFinishedRuns)
    });
  }

  getLabels() {
    return this.get('labels') || {};
  }

  getMem() {
    const {mem = MesosConstants.MIN_MEM} = this.get('run') || {};

    return mem;
  }

  getLastRunsSummary() {
    return this.get('historySummary') || {};
  }

  getLastRunStatus() {
    let {lastFailureAt, lastSuccessAt} = this.getLastRunsSummary();
    let status = 'N/A';
    let time = null;

    if (lastFailureAt !== null) {
      lastFailureAt = DateUtil.strToMs(lastFailureAt);
    }

    if (lastSuccessAt !== null) {
      lastSuccessAt = DateUtil.strToMs(lastSuccessAt);
    }

    if (lastFailureAt !== null || lastSuccessAt !== null) {
      if (lastFailureAt > lastSuccessAt) {
        status = 'Failed';
        time = lastFailureAt;
      } else {
        status = 'Success';
        time = lastSuccessAt;
      }
    }

    return {status, time};
  }

  getName() {
    return this.getId().split('.').pop();
  }

  getSchedules() {
    return this.get('schedules') || [];
  }

  getScheduleStatus() {
    let activeRuns = this.getActiveRuns();
    let activeRunsLength = activeRuns.getItems().length;
    let scheduleLength = this.getSchedules().length;

    if (activeRunsLength > 0) {
      let longestRunningActiveRun = activeRuns.getLongestRunningActiveRun();
      return longestRunningActiveRun.getStatus();
    }

    if (scheduleLength > 0) {
      let schedule = this.getSchedules()[0];

      if (!!schedule && schedule.enabled) {
        return 'SCHEDULED';
      }
    }

    if (scheduleLength === 0 && activeRunsLength === 0) {
      return 'UNSCHEDULED';
    }

    return 'COMPLETED';
  }

  toJSON() {
    return cleanJobJSON(this.get());
  }
};
