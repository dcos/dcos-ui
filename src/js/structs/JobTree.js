import Job from './Job';
import Tree from './Tree';

module.exports = class JobTree extends Tree {
  /**
   * (Chronos) JobTree
   * @param {{
   *          id:string,
   *          items:array<({id:string, items:array}|*)>,
   *          filterProperties:{propertyName:(null|string|function)}
   *        }} options
   * TODO: DCOS-7369 - Add remove item method
   * @constructor
   * @struct
   */
  constructor({id, items, filterProperties} = {}) {
    super({filterProperties});

    this.id = '/';

    if (typeof id == 'string') {
      if (id !== '/' && (!id.startsWith('/') || id.endsWith('/'))) {
        throw new Error(`Id (${id}) must start with a leading slash ("/") ` +
          'and should not end with a slash, except for root id which is only ' +
          'a slash.');
      }

      this.id = id;
    }

    // Parse and add items to the current tree
    if (Array.isArray(items)) {
      items.forEach((item)=> {
        // Create instances of JobTree for tree like items and add them
        if ((item.items != null && Array.isArray(item.items))
          && !(item instanceof JobTree)) {
          this.add(new this.constructor(
            Object.assign({filterProperties: this.getFilterProperties()}, item)
          ));

          return;
        }

        // Convert items into instance of Job and add them to the current tree
        if (!(item instanceof Job) && !(item instanceof JobTree)) {
          this.add(new Job(item));

          return;
        }

        this.add(item);
      })
    }
  }

  /**
   * Add item - This method will create sub trees if needed to insert the item
   * at the correct location (based on id/path matching).
   * @param {JobTree|Job} item
   */
  add(item) {
    if (!(item instanceof Job || item instanceof JobTree)) {
      throw new TypeError('item is neither an instance of Job nor JobTree');
    }

    const itemId = item.getId();

    if (!itemId.startsWith(this.getId())) {
      throw new Error(
        `item id (${itemId}) doesn't match tree id (${this.getId()})`
      );
    }

    // Check if the item is already present in the tree (including sub trees)
    if (this.findItemById(itemId)) {
      // TODO: DCOS-7370 - Update item (type Job) data
      return;
    }

    // Get the parent id (e.g. /group) by matching every thing but the item
    // name including the preceding slash "/" (e.g. /id).
    const [parentId] = itemId.match(/\/.*?(?=\/?[^/]+\/?$)/);

    // Add item to the current tree if it's the actual parent tree
    if (this.getId() === parentId) {
      super.add(item);
      return;
    }

    // Find or create corresponding parent tree and add it to the tree
    let parent = this.findItemById(parentId);
    if (!parent) {
      parent = new this.constructor({id: parentId});
      this.add(parent);
    }

    // Add item to parent tree
    parent.add(item);
  }

  getId() {
    return this.id;
  }

  /**
   * @param {string} id
   * @return {Job|JobTree} matching item
   */
  findItemById(id) {
    return this.findItem(function (item) {
      return item.getId() === id;
    });
  }

  getName() {
    return this.getId().split('/').pop();
  }
};
