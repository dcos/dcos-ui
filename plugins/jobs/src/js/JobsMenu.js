/* eslint-disable no-unused-vars */
import React from "react";
/* eslint-enable no-unused-vars */

import { componentFromStream } from "data-service";
import "rxjs/add/operator/combineLatest";

import Page from "#SRC/js/components/Page";
import JobsBreadcrumbs from "./components/JobsBreadcrumbs";
import { jobsRunNowAction, jobsRunNow$ } from "./JobsRunNow";

const getActions = function(props) {
  const [edit, ...rest] = props.actions;
  const newActions = [];

  newActions.push(edit);

  newActions.push(jobsRunNowAction(props.job.getId()));

  return newActions.concat(rest);
};

const JobsMenu = componentFromStream(props$ => {
  return props$.combineLatest(jobsRunNow$.startWith({}), props => {
    return (
      <Page.Header
        actions={getActions(props)}
        breadcrumbs={<JobsBreadcrumbs tree={props.tree} item={props.job} />}
        tabs={props.tabs}
        isPageHeader={props.isPageHeader}
      />
    );
  });
});

export default JobsMenu;
