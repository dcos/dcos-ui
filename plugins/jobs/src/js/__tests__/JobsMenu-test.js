/* eslint-disable no-unused-vars */
import React from "react";
/* eslint-enable no-unused-vars */

import Job from "#SRC/js/structs/Job";
import JobTree from "#SRC/js/structs/JobTree";

import Page from "#SRC/js/components/Page";
import JobsMenu from "../JobsMenu";

describe("LicensingExpirationRow", function() {
  describe("#render", function() {
    it("renders jobs menu", function() {
      const job = new Job({ id: "test" });
      const jobTree = new JobTree();

      const instance = (
        <Page>
          <JobsMenu
            actions={[""]}
            isPageHeader={true}
            job={job}
            tree={jobTree}
          />
          <div />
        </Page>
      );

      expect(instance).toMatchSnapshot();
    });
  });
});
