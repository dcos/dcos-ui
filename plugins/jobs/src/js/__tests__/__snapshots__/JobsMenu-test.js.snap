// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicensingExpirationRow #render renders jobs menu 1`] = `
<Page>
  <ComponentFromStream
    actions={
      Array [
        "",
      ]
    }
    isPageHeader={true}
    job={
      Object {
        "id": "test",
      }
    }
    tree={
      JobTree {
        "filterProperties": Object {},
        "id": "",
        "list": Array [],
      }
    }
  />
  <div />
</Page>
`;
